src/appleseed/foundation/math/knn/knn_anyquery.h

```diff
@@ -0,0 +1,397 @@
+
+        const VectorType parent_dvec = top_entry->m_dvec;
+
+        FOUNDATION_KNN_QUERY_STATS(++fetched_node_count);
+
+        while (node->is_interior())
+        {
+            const size_t split_dim = node->get_split_dim();
+            const ValueType split_abs = node->get_split_abs();
+            const ValueType split_dist = query_point[split_dim] - split_abs;
+
+            // Figure out which child node to follow (the one that contains the query point) and which one to push.
+            // Points on the split plane belong to the right child node.
+            const size_t follow_index = split_dist >= ValueType(0.0) ? 1 : 0;
+            const NodeType* APPLESEED_RESTRICT left_child_node = nodes + node->get_child_node_index();
+            const NodeType* APPLESEED_RESTRICT follow_node = left_child_node + follow_index;
+            const NodeType* APPLESEED_RESTRICT queue_node = left_child_node + 1 - follow_index;
+
+            FOUNDATION_KNN_QUERY_STATS(++fetched_node_count);
+
+            // Try to target nodes with an "ideal" number of points.
+            if (follow_node->get_point_count() < IdealLeafSize)
+                break;
+
+            // Push the node that we don't visit now to the node stack.
+            if (square(split_dist) < query_max_square_distance)
+            {
+                VectorType dvec = parent_dvec;
+                dvec[split_dim] = split_dist;
+
+                // Push the node to the node stack.
+                assert(node_queue_size < NodeQueueSize);
+                node_queue[node_queue_size++] = NodeEntry(queue_node, dvec, square_norm(dvec));
+
+                // Order the top levels of the node stack.
+                if (node_queue_size >= 4)
+                {
+                    ORDER_NODE_ENTRIES(4, 3);
+                    ORDER_NODE_ENTRIES(2, 1);
+                    ORDER_NODE_ENTRIES(4, 2);
+                    ORDER_NODE_ENTRIES(3, 1);
+                    ORDER_NODE_ENTRIES(3, 2);
+                }
+            }
+
+            // Continue with the child node.
+            node = follow_node;
+        }
+
+        FOUNDATION_KNN_QUERY_STATS(++visited_leaf_count);
+
+        const VectorType* APPLESEED_RESTRICT point_ptr = points + node->get_point_index();
+        const VectorType* APPLESEED_RESTRICT point_end = point_ptr + node->get_point_count();
+
+        while (point_ptr < point_end)
+        {
+            FOUNDATION_KNN_QUERY_STATS(++tested_point_count);
+
+            const ValueType square_dist = square_distance(*point_ptr++, query_point);
+
+            if (square_dist <= query_max_square_distance)
+                goto found;
+        }
+    }
+
+#undef ORDER_NODE_ENTRIES
+
+    FOUNDATION_KNN_QUERY_STATS(stats.m_fetched_nodes.insert(fetched_node_count));
+    FOUNDATION_KNN_QUERY_STATS(stats.m_visited_leaves.insert(visited_leaf_count));
+    FOUNDATION_KNN_QUERY_STATS(stats.m_tested_points.insert(tested_point_count));
+
+    return false;
+
+  found:
```

Comment:
Addition to the `goto` statement. This is badly readable, maybe rewrite it into an inlined function.